### ISSUE TITLE
Update Dataset Creation Form For Dynamic Fields

### ISSUE THREAD

**Issue by robertbartel** [NOAA-OWP/DMOD#198]
The [dataset management GUI view](https://github.com/robertbartel/DMOD/blob/e21/dataset_mgr_gui/main/python/gui/MaaS/cbv/DatasetManagementView.py) (still under review #197) has a component view and form for creating new datasets. At the moment, the dynamic field for an upload input is partially broke (although that may need to be dynamic). This needs to be corrected. 

It also does not yet have anything for defining the [DataDomain](https://github.com/robertbartel/DMOD/blob/b52cfd7f03f56b4100b2632620cdb33e49a26e39/python/lib/core/dmod/core/meta_data.py#L423) of the dataset. There should be some dynamic fields for setting these during creation. They should depend on the [DataFormat](https://github.com/robertbartel/DMOD/blob/b52cfd7f03f56b4100b2632620cdb33e49a26e39/python/lib/core/dmod/core/meta_data.py#L35) selection, as this will determine the [indices](https://github.com/robertbartel/DMOD/blob/b52cfd7f03f56b4100b2632620cdb33e49a26e39/python/lib/core/dmod/core/meta_data.py#L9) applicable within the domain.

**Comment by aaraney** [comment]
@robertbartel, looking at the source for [`createDatasetForm.js`](https://github.com/robertbartel/DMOD/blob/b52cfd7f03f56b4100b2632620cdb33e49a26e39/python/gui/MaaS/static/maas/js/components/createDatasetForm.js#L16), I see that `NETCDF_FORCING_CANONICAL` should have an upload input option, what other data formats, if any, should have an upload input option?

**Comment by robertbartel** [comment]
I think that particular input needs to be changed to apply to all the formats.  It may be worth having a checkbox or something to toggle whether it is shown.  The one exception could be ``OUTPUT``, but I think I am going to need to change that to be something more specific, and even then we may want it possible to upload already created results.

The things that really should be more dynamic are inputs to control the continuous and discrete restriction values; e.g., time range, the set of catchments, etc.  Exactly what those is specific to each format.

**Comment by aaraney** [comment]
> The one exception could be OUTPUT, but I think I am going to need to change that to be something more specific, and even then we may want it possible to upload already created results.

Thanks for clarifying. I am leaning towards leaving upload open to all data formats at the moment. My thinking is one person's output is or could be another person's input. It might make more sense for me to leave output off for now and just leave a simple way to change this in the future in the form of comments when we fully support such a feature.

> The things that really should be more dynamic are inputs to control the continuous and discrete restriction values; e.g., time range, the set of catchments, etc. Exactly what those is specific to each format.

I am not seeing those fields currently. Should I be?

**Comment by aaraney** [comment]
I think dynamic field removal issue is because of how we are removing children elements of `#dataset-create-form-div-dynamic-inputs`:  [`dy_div.childNodes.forEach(child => child.remove());`](https://github.com/robertbartel/DMOD/blob/b52cfd7f03f56b4100b2632620cdb33e49a26e39/python/gui/MaaS/static/maas/js/components/createDatasetForm.js#L13). `forEach` iterates and mutates that underlying vector, so when the first child is removed that is causing issues. Changing the code in `createDatasetForm.js` to the following should fix this issue:

```javascript
// remove all dynamic input child elements
while (dy_div.firstChild) {
  dy_div.removeChild(dy_div.firstChild);
}
```

**Comment by robertbartel** [comment]
> My thinking is one person's output is or could be another person's input.

Exactly.   This is why I am strongly considering removing the category `OUTPUT`, likely replacing it with several new categories (e.g., `NGEN_RESULTS`, etc.).  It becomes even more necessary if we have MaaS Job subtype(s) for deriving a forcing dataset from other existing forcing data. 

> I am not seeing those fields currently. Should I be?

No, not yet.  I got stuck at getting the dynamic logic working at all for something, and didn't start the others.  Feel free to set those up however seems most appropriate.

**Comment by aaraney** [comment]
@robertbartel, I think I am moving in the right direction, I still need to iron out the details of handling datetime objects, but before I move ahead below is what I have working at the moment. Ill get this into a PR this morning, but for discussion sake thought it made sense to add it here first.

<img width="1788" alt="image" src="https://user-images.githubusercontent.com/39420640/198300568-95e192af-705a-4f45-8698-1edbc9f78705.png">

**Comment by robertbartel** [comment]
First, there doesn't need to be an input for discrete versus continuous. The indices for which we want to input values are themselves intrinsically one or the other. This should inform the type of input component that gets used. 

Second, I don't think those are the indices for which we want values input. Values within the dataset, perhaps. But not domain defining properties. I.e., the user won't need it determined whether datasets they need are available by looking at the range of rain rate values. The times and catchments for which data is available are more likely the things to check for something like forcings. But those things are explicitly part of each DataFormat enum value.

**Comment by aaraney** [comment]
Okay, that makes sense. And sorry for what in retrospect are pretty naive assumptions, I thought it was a little odd to include a toggle for discrete vs continuous given we typically know the difference _a priori_. In reading the commented out sections of `dataset_management.html` in your `e21/dataset_mgr_gui/main`, it wasn't clear to me if there should or shouldn't be a toggle between discrete and continuous. 

Okay, thanks for clarifying regarding the input values. I think I was thrown off by the use of `parameter` in `dataset_management.html`. I took that to mean a `DataFormat`'s `data_fields`. Admittedly, I thought it was a little odd to get min max values for each _data field_, but alas here we are.

Luckily these shouldn't be huge changes and I think we are on the same page and moving in the right direction now.

**Comment by robertbartel** [comment]
Now closed by #240.

**Comment by robertbartel** [comment]
The original associated PR for this was closed without merging, with (some, but not all of the) various pieces of work separated into individual PRs.  As such, this needs to be moved back to another status.  

For now, I am moving to _Requested_, since it is realistic we are going to re-implement the entire dataset management GUI view.
